pfs_middleware/pfs_middleware/middleware.py

```diff
@@ -1200,8 +1205,16 @@ def _json_container_get_response(self, container_entries, account_name):
                 "hash": etag,
                 "last_modified": last_modified}
             json_entries.append(json_entry)
+
+            if delimiter != "" and "IsDir" in ent and ent["IsDir"]:
+                json_entries.append({"subdir": ent["Basename"] + delimiter})
+
         return json.dumps(json_entries)
 
+    # TODO: This method is usually non reachable, because at some point in the
+    # pipeline, we convert JSON to XML. We should either remove this or update
+    # it to support delimiters in case it's really needed.
+    # Same thing probably applies to plain text responses.
     def _xml_container_get_response(self, container_entries, account_name,
```

Comment:
Doesn't this get called if the request specifies XML response?

Reply:
Depends on the version of Swift -- unfortunately. With Swift 2.16.0 or later (and 2.15.3 SwiftStack Swift release), this middleware will only ever see content type as application/json. There is a different piece of middleware -- `listing_formats`, which requests JSON back and then mutates it to any form that the client requests (plain or xml). We can probably remove this additional formatting code, actually, and always return JSON if we have a guarantee about the Swift version.